File: python_tests/page_objects/administration_page.py
"""
Page Object Model for Administration Prototype Page
Provides methods to interact with UI elements in the admin-prototype.html page
"""
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import Select
from selenium.webdriver.common.keys import Keys
from selenium.common.exceptions import TimeoutException, NoSuchElementException
import time
from config import TestConfig

class AdministrationPage:
    """Page Object for Administration prototype page"""
    
    def __init__(self, driver):
        self.driver = driver
        self.wait = WebDriverWait(driver, TestConfig.IMPLICIT_WAIT)
    
    # ===== NAVIGATION SELECTORS =====
    NAV_USER_MANAGEMENT = (By.ID, "nav-users")
    NAV_ROLE_MANAGEMENT = (By.ID, "nav-roles")  
    NAV_SETTINGS = (By.ID, "nav-settings")
    
    # ===== USER MANAGEMENT SELECTORS =====
    # User table
    USER_TABLE = (By.ID, "userTable")
    USER_TABLE_ROWS = (By.CSS_SELECTOR, "#userTable tbody tr")
    USER_TABLE_HEADERS = (By.CSS_SELECTOR, "#userTable thead th")
    
    # User actions
    ADD_USER_BTN = (By.ID, "addUserBtn")
    BULK_UPLOAD_BTN = (By.ID, "bulkUploadBtn")
    DELETE_SELECTED_BTN = (By.ID, "deleteSelectedBtn")
    
    # User form fields
    USER_ID_INPUT = (By.ID, "userId")
    USER_NAME_INPUT = (By.ID, "userName")
    USER_EMAIL_INPUT = (By.ID, "userEmail")
    USER_DEPARTMENT_SELECT = (By.ID, "userDepartment")
    USER_ROLE_SELECT = (By.ID, "userRole")
    USER_STATUS_SELECT = (By.ID, "userStatus")
    USER_LOCATION_SELECT = (By.ID, "userLocation")
    
    # User form buttons
    SAVE_USER_BTN = (By.ID, "saveUserBtn")
    CANCEL_USER_BTN = (By.ID, "cancelUserBtn")
    
    # Search and filters
    USER_SEARCH_INPUT = (By.ID, "userSearch")
    DEPARTMENT_FILTER = (By.ID, "departmentFilter")
    ROLE_FILTER = (By.ID, "roleFilter")
    STATUS_FILTER = (By.ID, "statusFilter")
    LOCATION_FILTER = (By.ID, "locationFilter")
    CLEAR_FILTERS_BTN = (By.ID, "clearFiltersBtn")
    
    # ===== ROLE MANAGEMENT SELECTORS =====
    # Role table
    ROLE_TABLE = (By.ID, "roleTable")
    ROLE_TABLE_ROWS = (By.CSS_SELECTOR, "#roleTable tbody tr")
    
    # Role actions
    ADD_ROLE_BTN = (By.ID, "addRoleBtn")
    EDIT_ROLE_BTN = (By.CSS_SELECTOR, ".edit-role-btn")
    DELETE_ROLE_BTN = (By.CSS_SELECTOR, ".delete-role-btn")
    
    # Role form fields
    ROLE_NAME_INPUT = (By.ID, "roleName")
    ROLE_DESCRIPTION_INPUT = (By.ID, "roleDescription")
    PERMISSIONS_CHECKBOXES = (By.CSS_SELECTOR, "input[name='permissions']")
    
    # Role form buttons
    SAVE_ROLE_BTN = (By.ID, "saveRoleBtn")
    CANCEL_ROLE_BTN = (By.ID, "cancelRoleBtn")
    
    # ===== SETTINGS SELECTORS =====
    # Communication Settings
    EMAIL_NOTIFICATIONS_TOGGLE = (By.ID, "emailNotifications")
    SMS_NOTIFICATIONS_TOGGLE = (By.ID, "smsNotifications")
    PUSH_NOTIFICATIONS_TOGGLE = (By.ID, "pushNotifications")
    
    # Delay Settings
    PLANNED_OUTAGE_DELAY = (By.ID, "plannedOutageDelay")
    UNPLANNED_OUTAGE_DELAY = (By.ID, "unplannedOutageDelay")
    EMERGENCY_DELAY = (By.ID, "emergencyDelay")
    
    # Extreme Weather Mode
    EXTREME_WEATHER_TOGGLE = (By.ID, "extremeWeatherMode")
    WEATHER_THRESHOLD_TEMP = (By.ID, "weatherThresholdTemp")
    WEATHER_THRESHOLD_WIND = (By.ID, "weatherThresholdWind")
    
    # Settings buttons
    SAVE_SETTINGS_BTN = (By.ID, "saveSettingsBtn")
    RESET_SETTINGS_BTN = (By.ID, "resetSettingsBtn")
    
    # ===== COMMON SELECTORS =====
    SUCCESS_MESSAGE = (By.CSS_SELECTOR, ".alert-success")
    ERROR_MESSAGE = (By.CSS_SELECTOR, ".alert-danger")
    CONFIRMATION_MODAL = (By.ID, "confirmationModal")
    CONFIRM_YES_BTN = (By.ID, "confirmYes")
    CONFIRM_NO_BTN = (By.ID, "confirmNo")
    
    def slow_action(self, seconds=None):
        """Add delay between actions for visibility"""
        if TestConfig.SLOW_MOTION:
            delay = seconds if seconds else TestConfig.SLOW_MOTION_DELAY
            time.sleep(delay)
    
    # ===== NAVIGATION METHODS =====
    def click_user_management_nav(self):
        """Click on User Management navigation"""
        self.wait.until(EC.element_to_be_clickable(self.NAV_USER_MANAGEMENT)).click()
        self.slow_action()
        return self
    
    def click_role_management_nav(self):
        """Click on Role Management navigation"""
        self.wait.until(EC.element_to_be_clickable(self.NAV_ROLE_MANAGEMENT)).click()
        self.slow_action()
        return self
    
    def click_settings_nav(self):
        """Click on Settings navigation"""
        self.wait.until(EC.element_to_be_clickable(self.NAV_SETTINGS)).click()
        self.slow_action()
        return self
    
    # ===== USER MANAGEMENT METHODS =====
    def get_user_table_headers(self):
        """Get all table headers from user table"""
        headers = self.driver.find_elements(*self.USER_TABLE_HEADERS)
        return [header.text for header in headers]
    
    def get_user_table_rows_count(self):
        """Get number of rows in user table"""
        try:
            rows = self.driver.find_elements(*self.USER_TABLE_ROWS)
            return len(rows)
        except NoSuchElementException:
            return 0
    
    def get_user_table_data(self):
        """Get all user data from table"""
        users = []
        try:
            rows = self.driver.find_elements(*self.USER_TABLE_ROWS)
            for row in rows:
                cells = row.find_elements(By.TAG_NAME, "td")
                if len(cells) >= 6:  # Assuming 6+ columns
                    user_data = {
                        'id': cells[1].text,  # Skip checkbox column
                        'name': cells[2].text,
                        'email': cells[3].text,
                        'department': cells[4].text,
                        'role': cells[5].text,
                        'status': cells[6].text if len(cells) > 6 else ''
                    }
                    users.append(user_data)
        except NoSuchElementException:
            pass
        return users
    
    def click_add_user_button(self):
        """Click Add User button"""
        self.wait.until(EC.element_to_be_clickable(self.ADD_USER_BTN)).click()
        self.slow_action()
        return self
    
    def fill_user_form(self, user_id=None, name=None, email=None, department=None, role=None, status=None, location=None):
        """Fill user form fields"""
        if user_id:
            self.wait.until(EC.presence_of_element_located(self.USER_ID_INPUT)).clear()
            self.driver.find_element(*self.USER_ID_INPUT).send_keys(user_id)
            self.slow_action(0.2)
        
        if name:
            self.driver.find_element(*self.USER_NAME_INPUT).clear()
            self.driver.find_element(*self.USER_NAME_INPUT).send_keys(name)
            self.slow_action(0.2)
        
        if email:
            self.driver.find_element(*self.USER_EMAIL_INPUT).clear()
            self.driver.find_element(*self.USER_EMAIL_INPUT).send_keys(email)
            self.slow_action(0.2)
        
        if department:
            Select(self.driver.find_element(*self.USER_DEPARTMENT_SELECT)).select_by_visible_text(department)
            self.slow_action(0.2)
        
        if role:
            Select(self.driver.find_element(*self.USER_ROLE_SELECT)).select_by_visible_text(role)
            self.slow_action(0.2)
        
        if status:
            Select(self.driver.find_element(*self.USER_STATUS_SELECT)).select_by_visible_text(status)
            self.slow_action(0.2)
        
        if location:
            Select(self.driver.find_element(*self.USER_LOCATION_SELECT)).select_by_visible_text(location)
            self.slow_action(0.2)
        
        return self
    
    def save_user(self):
        """Click Save User button"""
        self.wait.until(EC.element_to_be_clickable(self.SAVE_USER_BTN)).click()
        self.slow_action()
        return self
    
    def cancel_user(self):
        """Click Cancel User button"""
        self.wait.until(EC.element_to_be_clickable(self.CANCEL_USER_BTN)).click()
        self.slow_action()
        return self
    
    def search_users(self, search_term):
        """Search users using search input"""
        search_input = self.wait.until(EC.presence_of_element_located(self.USER_SEARCH_INPUT))
        search_input.clear()
        search_input.send_keys(search_term)
        search_input.send_keys(Keys.ENTER)
        self.slow_action()
        return self
    
    def filter_users_by_department(self, department):
        """Filter users by department"""
        Select(self.driver.find_element(*self.DEPARTMENT_FILTER)).select_by_visible_text(department)
        self.slow_action()
        return self
    
    def filter_users_by_role(self, role):
        """Filter users by role"""
        Select(self.driver.find_element(*self.ROLE_FILTER)).select_by_visible_text(role)
        self.slow_action()
        return self
    
    def filter_users_by_status(self, status):
        """Filter users by status"""
        Select(self.driver.find_element(*self.STATUS_FILTER)).select_by_visible_text(status)
        self.slow_action()
        return self
    
    def clear_all_filters(self):
        """Click clear filters button"""
        self.wait.until(EC.element_to_be_clickable(self.CLEAR_FILTERS_BTN)).click()
        self.slow_action()
        return self
    
    def select_user_checkbox(self, user_index):
        """Select user checkbox by index (0-based)"""
        rows = self.driver.find_elements(*self.USER_TABLE_ROWS)
        if user_index < len(rows):
            checkbox = rows[user_index].find_element(By.CSS_SELECTOR, "input[type='checkbox']")
            checkbox.click()
            self.slow_action()
        return self
    
    def click_bulk_upload_button(self):
        """Click Bulk Upload button"""
        self.wait.until(EC.element_to_be_clickable(self.BULK_UPLOAD_BTN)).click()
        self.slow_action()
        return self
    
    def click_delete_selected_button(self):
        """Click Delete Selected button"""
        self.wait.until(EC.element_to_be_clickable(self.DELETE_SELECTED_BTN)).click()
        self.slow_action()
        return self
    
    # ===== ROLE MANAGEMENT METHODS =====
    def get_role_table_rows_count(self):
        """Get number of rows in role table"""
        try:
            rows = self.driver.find_elements(*self.ROLE_TABLE_ROWS)
            return len(rows)
        except NoSuchElementException:
            return 0
    
    def click_add_role_button(self):
        """Click Add Role button"""
        self.wait.until(EC.element_to_be_clickable(self.ADD_ROLE_BTN)).click()
        self.slow_action()
        return self
    
    def fill_role_form(self, role_name=None, description=None, permissions=None):
        """Fill role form fields"""
        if role_name:
            self.wait.until(EC.presence_of_element_located(self.ROLE_NAME_INPUT)).clear()
            self.driver.find_element(*self.ROLE_NAME_INPUT).send_keys(role_name)
            self.slow_action(0.2)
        
        if description:
            self.driver.find_element(*self.ROLE_DESCRIPTION_INPUT).clear()
            self.driver.find_element(*self.ROLE_DESCRIPTION_INPUT).send_keys(description)
            self.slow_action(0.2)
        
        if permissions:
            checkboxes = self.driver.find_elements(*self.PERMISSIONS_CHECKBOXES)
            for checkbox in checkboxes:
                permission_name = checkbox.get_attribute('value')
                if permission_name in permissions:
                    if not checkbox.is_selected():
                        checkbox.click()
                        self.slow_action(0.1)
                else:
                    if checkbox.is_selected():
                        checkbox.click()
                        self.slow_action(0.1)
        
        return self
    
    def save_role(self):
        """Click Save Role button"""
        self.wait.until(EC.element_to_be_clickable(self.SAVE_ROLE_BTN)).click()
        self.slow_action()
        return self
    
    # ===== SETTINGS METHODS =====
    def toggle_email_notifications(self, enable=True):
        """Toggle email notifications setting"""
        toggle = self.wait.until(EC.element_to_be_clickable(self.EMAIL_NOTIFICATIONS_TOGGLE))
        if toggle.is_selected() != enable:
            toggle.click()
            self.slow_action()
        return self
    
    def toggle_sms_notifications(self, enable=True):
        """Toggle SMS notifications setting"""
        toggle = self.wait.until(EC.element_to_be_clickable(self.SMS_NOTIFICATIONS_TOGGLE))
        if toggle.is_selected() != enable:
            toggle.click()
            self.slow_action()
        return self
    
    def set_planned_outage_delay(self, minutes):
        """Set planned outage delay"""
        delay_input = self.wait.until(EC.presence_of_element_located(self.PLANNED_OUTAGE_DELAY))
        delay_input.clear()
        delay_input.send_keys(str(minutes))
        self.slow_action()
        return self
    
    def toggle_extreme_weather_mode(self, enable=True):
        """Toggle extreme weather mode"""
        toggle = self.wait.until(EC.element_to_be_clickable(self.EXTREME_WEATHER_TOGGLE))
        if toggle.is_selected() != enable:
            toggle.click()
            self.slow_action()
        return self
    
    def save_settings(self):
        """Click Save Settings button"""
        self.wait.until(EC.element_to_be_clickable(self.SAVE_SETTINGS_BTN)).click()
        self.slow_action()
        return self
    
    def reset_settings(self):
        """Click Reset Settings button"""
        self.wait.until(EC.element_to_be_clickable(self.RESET_SETTINGS_BTN)).click()
        self.slow_action()
        return self
    
    # ===== COMMON METHODS =====
    def wait_for_success_message(self, timeout=10):
        """Wait for success message to appear"""
        try:
            element = WebDriverWait(self.driver, timeout).until(
                EC.presence_of_element_located(self.SUCCESS_MESSAGE)
            )
            return element.text
        except TimeoutException:
            return None
    
    def wait_for_error_message(self, timeout=10):
        """Wait for error message to appear"""
        try:
            element = WebDriverWait(self.driver, timeout).until(
                EC.presence_of_element_located(self.ERROR_MESSAGE)
            )
            return element.text
        except TimeoutException:
            return None
    
    def confirm_action(self, confirm=True):
        """Handle confirmation modal"""
        try:
            self.wait.until(EC.presence_of_element_located(self.CONFIRMATION_MODAL))
            if confirm:
                self.wait.until(EC.element_to_be_clickable(self.CONFIRM_YES_BTN)).click()
            else:
                self.wait.until(EC.element_to_be_clickable(self.CONFIRM_NO_BTN)).click()
            self.slow_action()
        except TimeoutException:
            pass  # No confirmation modal appeared
        return self
    
    def is_element_present(self, locator, timeout=5):
        """Check if element is present on page"""
        try:
            WebDriverWait(self.driver, timeout).until(EC.presence_of_element_located(locator))
            return True
        except TimeoutException:
            return False
    
    def is_element_visible(self, locator, timeout=5):
        """Check if element is visible on page"""
        try:
            WebDriverWait(self.driver, timeout).until(EC.visibility_of_element_located(locator))
            return True
        except TimeoutException:
            return False

File: playwright_tests/page-objects/administration-page.js
/**
 * Page Object Model for Administration Prototype Page
 * Provides methods to interact with UI elements in the admin-prototype.html page
 * JavaScript port of python_tests/page_objects/administration_page.py
 */

class AdministrationPage {
  constructor(page) {
    this.page = page;
    this.slowMo = 500; // milliseconds, matching Playwright config
    
    // ===== NAVIGATION SELECTORS =====
    this.navUserManagement = 'a.tab:has-text("Users")';
    this.navRoleManagement = 'a.tab:has-text("Roles")';
    this.navSettings = 'a.tab:has-text("Settings")';
    
    // ===== USER MANAGEMENT SELECTORS =====
    // User table
    this.userTable = '#userTable';
    this.userTableRows = '#userTable tbody tr';
    this.userTableHeaders = '#userTable thead th';
    
    // User actions
    this.addUserBtn = '#addUserBtn';
    this.bulkUploadBtn = '#bulkUploadBtn';
    this.deleteSelectedBtn = '#deleteSelectedBtn';
    
    // User form fields
    this.userIdInput = '#userId';
    this.userNameInput = '#userName';
    this.userEmailInput = '#userEmail';
    this.userRoleDropdown = '#userRole'; // Multi-select dropdown
    this.userStatusSelect = '#userStatus';
    
    // User form buttons
    this.saveUserBtn = '#saveUserBtn';
    this.cancelUserBtn = '#cancelUserBtn';
    
    // Search and filters
    this.userSearchInput = '#userSearch';
    this.roleFilter = '#roleFilter'; // Multi-select dropdown
    this.statusFilter = '#statusFilter';
    this.clearFiltersBtn = '#clearFiltersBtn';
    
    // ===== ROLE MANAGEMENT SELECTORS =====
    // Role table
    this.roleTable = '#roleTable';
    this.roleTableRows = '#roleTable tbody tr';
    
    // Role actions
    this.addRoleBtn = '#addRoleBtn';
    this.editRoleBtn = '.edit-role-btn';
    this.deleteRoleBtn = '.delete-role-btn';
    
    // Role form fields
    this.roleNameInput = '#roleName';
    this.roleDescriptionInput = '#roleDescription';
    this.permissionsCheckboxes = 'input[name="permissions"]';
    
    // Role form buttons
    this.saveRoleBtn = '#saveRoleBtn';
    this.cancelRoleBtn = '#cancelRoleBtn';
    
    // ===== SETTINGS SELECTORS =====
    // Communication Settings
    this.emailNotificationsToggle = '#emailNotifications';
    this.smsNotificationsToggle = '#smsNotifications';
    this.pushNotificationsToggle = '#pushNotifications';
    
    // Delay Settings
    this.plannedOutageDelay = '#plannedOutageDelay';
    this.unplannedOutageDelay = '#unplannedOutageDelay';
    this.emergencyDelay = '#emergencyDelay';
    
    // Extreme Weather Mode
    this.extremeWeatherToggle = '#extremeWeatherMode';
    this.weatherThresholdTemp = '#weatherThresholdTemp';
    this.weatherThresholdWind = '#weatherThresholdWind';
    
    // Settings buttons
    this.saveSettingsBtn = '#saveSettingsBtn';
    this.resetSettingsBtn = '#resetSettingsBtn';
    
    // ===== COMMON SELECTORS =====
    this.successMessage = '.alert-success';
    this.errorMessage = '.alert-danger';
    this.confirmationModal = '#confirmationModal';
    this.confirmYesBtn = '#confirmYes';
    this.confirmNoBtn = '#confirmNo';
  }

  /**
   * Add delay between actions for visibility (matches pytest slow motion)
   */
  async slowAction(milliseconds = null) {
    const delay = milliseconds || this.slowMo;
    await this.page.waitForTimeout(delay);
  }

  // ===== NAVIGATION METHODS =====
  
  async clickUserManagementNav() {
    await this.page.click(this.navUserManagement);
    await this.slowAction();
    return this;
  }

  async clickRoleManagementNav() {
    await this.page.click(this.navRoleManagement);
    await this.slowAction();
    return this;
  }

  async clickSettingsNav() {
    await this.page.click(this.navSettings);
    await this.slowAction();
    return this;
  }

  // ===== USER MANAGEMENT METHODS =====
  
  async getUserTableHeaders() {
    const headers = await this.page.locator(this.userTableHeaders).allTextContents();
    return headers;
  }

  async getUserTableRowsCount() {
    try {
      const rows = await this.page.locator(this.userTableRows).count();
      return rows;
    } catch (error) {
      return 0;
    }
  }

  async getUserTableData() {
    const users = [];
    const rows = await this.page.locator(this.userTableRows).all();
    
    for (const row of rows) {
      // Skip hidden rows (filtered out by filter functions)
      const isVisible = await row.isVisible();
      if (!isVisible) {
        continue;
      }
      
      const cells = await row.locator('td').all();
      // Table columns: [0] Checkbox, [1] Name, [2] Email, [3] Role, [4] Status, [5] Last Login, [6] Actions
      if (cells.length >= 6) {
        const userData = {
          name: (await cells[1].textContent()).trim(),
          email: (await cells[2].textContent()).trim(),
          role: (await cells[3].textContent()).trim(),
          status: (await cells[4].textContent()).trim(),
          lastLogin: cells.length > 5 ? (await cells[5].textContent()).trim() : ''
        };
        users.push(userData);
      }
    }
    
    return users;
  }

  async clickAddUserButton() {
    await this.page.click(this.addUserBtn);
    await this.slowAction();
    return this;
  }

  async fillUserForm({ userId, name, email, role }) {
    if (userId) {
      await this.page.fill(this.userIdInput, '');
      await this.page.fill(this.userIdInput, userId);
      await this.slowAction(200);
    }

    if (name) {
      await this.page.fill(this.userNameInput, '');
      await this.page.fill(this.userNameInput, name);
      await this.slowAction(200);
    }

    if (email) {
      await this.page.fill(this.userEmailInput, '');
      await this.page.fill(this.userEmailInput, email);
      await this.slowAction(200);
    }

    if (role) {
      // Handle multi-select role dropdown (new compact design)
      // First, open the dropdown by clicking the button
      await this.page.click(`${this.userRoleDropdown} .multi-select-button`);
      await this.slowAction(200);
      
      // role can be string or array
      const roles = Array.isArray(role) ? role : [role];
      
      // Click each role checkbox
      for (const roleName of roles) {
        const checkbox = await this.page.locator(`${this.userRoleDropdown} input[value="${roleName}"]`);
        const isChecked = await checkbox.isChecked();
        if (!isChecked) {
          await checkbox.click();
          await this.slowAction(100);
        }
      }
      
      // Close dropdown by clicking button again
      await this.page.click(`${this.userRoleDropdown} .multi-select-button`);
      await this.slowAction(200);
    }

    // Status is removed - new users are always Active

    return this;
  }

  async saveUser() {
    await this.page.click(this.saveUserBtn);
    await this.slowAction();
    return this;
  }

  async cancelUser() {
    await this.page.click(this.cancelUserBtn);
    await this.slowAction();
    return this;
  }

  async searchUsers(searchTerm) {
    await this.page.fill(this.userSearchInput, '');
    await this.page.fill(this.userSearchInput, searchTerm);
    await this.page.press(this.userSearchInput, 'Enter');
    await this.slowAction();
    return this;
  }

  async filterUsersByRole(role) {
    // Role filter is a multi-select dropdown
    // First, open the dropdown
    await this.page.click(`${this.roleFilter} .multi-select-button`);
    await this.slowAction(200);
    
    // Find and click the checkbox for the role
    const roleId = `role_filter_${role.toLowerCase().replace(/\s+/g, '_')}`;
    await this.page.click(`#${roleId}`);
    
    // Wait for the filter to be applied (onchange triggers filterUsers())
    await this.slowAction(500);
    
    // Close dropdown by clicking outside or button again
    await this.page.click(`${this.roleFilter} .multi-select-button`);
    await this.slowAction();
    return this;
  }

  async filterUsersByStatus(status) {
    await this.page.selectOption(this.statusFilter, status);
    await this.slowAction();
    return this;
  }

  async clearAllFilters() {
    await this.page.click(this.clearFiltersBtn);
    await this.slowAction();
    return this;
  }

  async selectUserCheckbox(userIndex) {
    const rows = await this.page.locator(this.userTableRows).all();
    if (userIndex < rows.length) {
      await rows[userIndex].locator('input[type="checkbox"]').click();
      await this.slowAction();
    }
    return this;
  }

  async clickBulkUploadButton() {
    await this.page.click(this.bulkUploadBtn);
    await this.slowAction();
    return this;
  }

  async clickDeleteSelectedButton() {
    await this.page.click(this.deleteSelectedBtn);
    await this.slowAction();
    return this;
  }

  // ===== ROLE MANAGEMENT METHODS =====
  
  async getRoleTableRowsCount() {
    try {
      const rows = await this.page.locator(this.roleTableRows).count();
      return rows;
    } catch (error) {
      return 0;
    }
  }

  async clickAddRoleButton() {
    await this.page.click(this.addRoleBtn);
    await this.slowAction();
    return this;
  }

  async fillRoleForm({ roleName, description, permissions }) {
    if (roleName) {
      await this.page.fill(this.roleNameInput, '');
      await this.page.fill(this.roleNameInput, roleName);
      await this.slowAction(200);
    }

    if (description) {
      await this.page.fill(this.roleDescriptionInput, '');
      await this.page.fill(this.roleDescriptionInput, description);
      await this.slowAction(200);
    }

    if (permissions) {
      // Get all permission items (the clickable divs)
      const permissionItems = await this.page.locator('.permission-item-modal').all();
      
      for (const item of permissionItems) {
        const permission = await item.getAttribute('data-permission');
        const isSelected = await item.evaluate(el => el.classList.contains('selected'));
        
        const shouldBeSelected = permissions.includes(permission);
        
        // Toggle if current state doesn't match desired state
        if (shouldBeSelected !== isSelected) {
          await item.click();
          await this.slowAction(100);
        }
      }
    }

    return this;
  }

  async saveRole() {
    await this.page.click(this.saveRoleBtn);
    await this.slowAction();
    return this;
  }

  async cancelRole() {
    await this.page.click(this.cancelRoleBtn);
    await this.slowAction();
    return this;
  }

  // ===== SETTINGS METHODS =====
  
  async expandSettingsGroup(groupTitle) {
    // Find the settings group by its title and expand it if collapsed
    const groupHeader = this.page.locator(`.settings-group-header:has-text("${groupTitle}")`);
    await groupHeader.waitFor({ state: 'visible', timeout: 5000 });
    
    // Check if the group is collapsed (icon is ▶)
    const icon = groupHeader.locator('.settings-group-icon');
    const iconText = await icon.textContent();
    
    if (iconText.includes('▶')) {
      // Group is collapsed, click to expand
      await groupHeader.click();
      await this.slowAction(500);
    }
    return this;
  }
  
  async toggleEmailNotifications(enable = true) {
    const toggle = this.page.locator(this.emailNotificationsToggle);
    await toggle.waitFor({ state: 'visible', timeout: 10000 });
    const isActive = await toggle.evaluate(el => el.classList.contains('active'));
    
    if (isActive !== enable) {
      await toggle.click();
      await this.slowAction(300);
      // Handle confirmation modal if it appears
      try {
        const okButton = this.page.locator('#communicationToggleModal button.btn-modal-primary');
        await okButton.waitFor({ state: 'visible', timeout: 2000 });
        await okButton.click();
        await this.slowAction();
      } catch (e) {
        // Modal didn't appear, continue
      }
    }
    return this;
  }

  async toggleSmsNotifications(enable = true) {
    const toggle = this.page.locator(this.smsNotificationsToggle);
    await toggle.waitFor({ state: 'visible', timeout: 10000 });
    const isActive = await toggle.evaluate(el => el.classList.contains('active'));
    
    if (isActive !== enable) {
      await toggle.click();
      await this.slowAction(300);
      // Handle confirmation modal if it appears  
      try {
        const okButton = this.page.locator('#communicationToggleModal button.btn-modal-primary');
        await okButton.waitFor({ state: 'visible', timeout: 2000 });
        await okButton.click();
        await this.slowAction();
      } catch (e) {
        // Modal didn't appear, continue
      }
    }
    return this;
  }

  async togglePushNotifications(enable = true) {
    const toggle = this.page.locator(this.pushNotificationsToggle);
    await toggle.waitFor({ state: 'visible', timeout: 10000 });
    const isActive = await toggle.evaluate(el => el.classList.contains('active'));
    
    if (isActive !== enable) {
      await toggle.click();
      await this.slowAction();
    }
    return this;
  }

  async setPlannedOutageDelay(minutes) {
    // Expand Delay & Control group if needed
    await this.expandSettingsGroup('Delay & Control');
    await this.page.fill(this.plannedOutageDelay, '');
    await this.page.fill(this.plannedOutageDelay, String(minutes));
    await this.slowAction();
    return this;
  }

  async setUnplannedOutageDelay(minutes) {
    // Expand Delay & Control group if needed
    await this.expandSettingsGroup('Delay & Control');
    await this.page.fill(this.unplannedOutageDelay, '');
    await this.page.fill(this.unplannedOutageDelay, String(minutes));
    await this.slowAction();
    return this;
  }

  async setEmergencyDelay(minutes) {
    // Expand Delay & Control group if needed
    await this.expandSettingsGroup('Delay & Control');
    await this.page.fill(this.emergencyDelay, '');
    await this.page.fill(this.emergencyDelay, String(minutes));
    await this.slowAction();
    return this;
  }

  async toggleExtremeWeatherMode(enable = true) {
    // First expand the Trigger Mode settings group
    await this.expandSettingsGroup('Trigger Mode');
    
    const toggle = this.page.locator(this.extremeWeatherToggle);
    await toggle.waitFor({ state: 'visible', timeout: 10000 });
    const isActive = await toggle.evaluate(el => el.classList.contains('active'));
    
    if (isActive !== enable) {
      await toggle.click();
      await this.slowAction();
      // Handle confirmation modal if it appears
      try {
        const okButton = this.page.locator('#extremeWeatherModal button.btn-modal-primary');
        await okButton.waitFor({ state: 'visible', timeout: 2000 });
        await okButton.click();
        await this.slowAction();
      } catch (e) {
        // Modal didn't appear, continue
      }
    }
    return this;
  }

  async setWeatherThresholdTemp(celsius) {
    // Expand Trigger Mode group if needed
    await this.expandSettingsGroup('Trigger Mode');
    await this.page.fill(this.weatherThresholdTemp, '');
    await this.page.fill(this.weatherThresholdTemp, String(celsius));
    await this.slowAction();
    return this;
  }

  async setWeatherThresholdWind(kmh) {
    // Expand Trigger Mode group if needed
    await this.expandSettingsGroup('Trigger Mode');
    await this.page.fill(this.weatherThresholdWind, '');
    await this.page.fill(this.weatherThresholdWind, String(kmh));
    await this.slowAction();
    return this;
  }

  async saveSettings() {
    await this.page.click(this.saveSettingsBtn);
    await this.slowAction();
    return this;
  }

  async resetSettings() {
    await this.page.click(this.resetSettingsBtn);
    await this.slowAction();
    return this;
  }

  // ===== COMMON METHODS =====
  
  async waitForSuccessMessage(timeout = 10000) {
    try {
      const message = await this.page.locator(this.successMessage).waitFor({ timeout });
      return await message.textContent();
    } catch (error) {
      return null;
    }
  }

  async waitForErrorMessage(timeout = 10000) {
    try {
      const message = await this.page.locator(this.errorMessage).waitFor({ timeout });
      return await message.textContent();
    } catch (error) {
      return null;
    }
  }

  async confirmAction(confirm = true) {
    try {
      await this.page.locator(this.confirmationModal).waitFor({ timeout: 5000 });
      
      if (confirm) {
        await this.page.click(this.confirmYesBtn);
      } else {
        await this.page.click(this.confirmNoBtn);
      }
      
      await this.slowAction();
    } catch (error) {
      // No confirmation modal appeared
    }
    return this;
  }

  async isElementPresent(selector, timeout = 5000) {
    try {
      await this.page.locator(selector).waitFor({ timeout });
      return true;
    } catch (error) {
      return false;
    }
  }

  async isElementVisible(selector, timeout = 5000) {
    try {
      await this.page.locator(selector).waitFor({ state: 'visible', timeout });
      return true;
    } catch (error) {
      return false;
    }
  }
}

module.exports = AdministrationPage;
